src/edu/washington/escience/myria/operator/Limit.java

```diff
@@ -38,23 +41,30 @@ public Limit(@Nonnull final Long limit, final Operator child) {
     this.limit = Objects.requireNonNull(limit, "limit");
     Preconditions.checkArgument(limit >= 0L, "limit must be non-negative");
     toEmit = this.limit;
+    done = false;
   }
 
   @Override
   protected TupleBatch fetchNextReady() throws DbException {
     Operator child = getChild();
-    for (TupleBatch tb = child.nextReady(); tb != null; tb = child.nextReady()) {
-      if (tb.numTuples() <= toEmit) {
-        toEmit -= tb.numTuples();
-        return tb;
-      } else if (toEmit > 0) {
-        tb = tb.prefix(Ints.checkedCast(toEmit));
-        toEmit = 0;
-        return tb;
+    if (done) {
```

Comment:
Replace this check with `child.isOpen()` or `child.eos()`. Actually, I think that this check is strictly unnecessary since after we've closed the child, its EOS status will be propagated to this operator and `nextReady()` will return `null` immediately. @jingjingwang could you comment?